library/bignum_mod.c

```diff
@@ -0,0 +1,153 @@
+/**
+ *  Internal bignum functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+
+#include <string.h>
+
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+#include "mbedtls/bignum.h"
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_printf      printf
+#define mbedtls_calloc      calloc
+#define mbedtls_free        free
+#endif
+
+#include "bignum_core.h"
+#include "bignum_mod.h"
+#include "bignum_mod_raw.h"
+#include "constant_time_internal.h"
+
+int mbedtls_mpi_mod_residue_setup( mbedtls_mpi_mod_residue *r,
+                                   mbedtls_mpi_mod_modulus *m,
+                                   mbedtls_mpi_uint *p,
+                                   size_t pn )
+{
+    if( p == NULL || m == NULL || r == NULL )
+        return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
+
+    if( pn < m->n || !mbedtls_mpi_core_lt_ct( m->p, p, pn ) )
+        return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
+
+    r->n = m->n;
+    r->p = p;
+
+    return( 0 );
+}
+
+void mbedtls_mpi_mod_residue_release( mbedtls_mpi_mod_residue *r )
+{
+    if ( r == NULL )
+        return;
+
+    r->n = 0;
+    r->p = NULL;
+}
+
+void mbedtls_mpi_mod_modulus_init( mbedtls_mpi_mod_modulus *m )
+{
+    if ( m == NULL )
+        return;
+
+    m->p = NULL;
+    m->n = 0;
+    m->plen = 0;
+    m->ext_rep = MBEDTLS_MPI_MOD_EXT_REP_INVALID;
+    m->int_rep = MBEDTLS_MPI_MOD_REP_INVALID;
+}
+
+void mbedtls_mpi_mod_modulus_free( mbedtls_mpi_mod_modulus *m )
+{
+    if ( m == NULL )
+        return;
+
+    switch( m->int_rep )
```

Comment:
return to my comment about "multiple statements on a single line", and also the discuss about enums, would this be better as

```
    switch( m->int_rep )
    {
        case MBEDTLS_MPI_MOD_REP_MONTGOMERY: mbedtls_free( m->rep.mont ); break;
        case MBEDTLS_MPI_MOD_REP_OPT_RED:    mbedtls_free( m->rep.ored ); break;
        case MBEDTLS_MPI_MOD_REP_INVALID:    break;
    }
```

Putting each case on one line is actually readable.

Not using `default` means that if we add a value to the enum then there are compiler flags that will tell us we are missing an enum value.

Reply:
I am not certain on the benefit of the proposed formatting. On top of making the code harder to read for some, it implies that each case will always contain a single statement. That will may not the case when `rep.mont` and `rep.ored` are populated. 

Also relying on compiler behaviour being consistent could suprise us. Different compilers or versions could change that particular flag's behavior. Whereas `default:¬ is the standard for decades, and most tools are tuned in to catch it.

Reply:
The suggestion was just a suggestion - for me, it is clearer to see what differs between the two cases, and in previous projects there have definitely been problems with code that were clear to see when aligned like this. And yes, it only works because of the short statements needed, but why not take advantage of properties of the code to make it more readable?

In terms of compiler behaviour, however, I disagree: we only need to use one compiler that tells us there's a problem with the source code. And yes, there are disagreements on the benefits of using `default` - while some secure coding standards require them, it can (and has done) hide problems that are introduced in later development of the code.

Reply:
I find the one line one statement version more readable too. I marked this addressed, but I didn't use the formatting, just removed the default.

As long as we don't have tools that catch a missing default, I think we should take advantage this compiler warning. Also, please note the we only use this if the default option would be a noop. If we throw an error by default, it remains, and most cases are like that.